File: backend/src/main/java/com/deck/lab/backend/model/CardAttribute.java
package com.deck.lab.backend.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

/**
 * Enum defining the elemental attributes of monster cards.
 *
 * <p>
 * <strong>Serialization Details:</strong>
 * </p>
 * <ul>
 * <li>{@code @JsonValue}: Serializes elements as capital string tokens (e.g.
 * "LIGHT", "DARK") conforming to database structures.</li>
 * <li>{@code @JsonCreator}: Parser enabling case-insensitive parameter
 * conversions from inbound requests.</li>
 * </ul>
 */
public enum CardAttribute {
    LIGHT("LIGHT"),
    DARK("DARK"),
    WATER("WATER"),
    FIRE("FIRE"),
    EARTH("EARTH"),
    WIND("WIND"),
    DIVINE("DIVINE");

    private final String value;

    CardAttribute(String value) {
        this.value = value;
    }

    /**
     * Gets the string representation of the card attribute.
     *
     * @return the serialized card attribute string
     */
    @JsonValue
    public String getValue() {
        return value;
    }

    /**
     * Resolves a card attribute from its string representation (case-insensitive).
     *
     * @param value the card attribute string
     * @return the resolved CardAttribute enum
     * @throws IllegalArgumentException if the string is unknown
     */
    @JsonCreator
    public static CardAttribute fromString(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        String trimmed = value.trim();
        for (CardAttribute attr : CardAttribute.values()) {
            if (attr.value.equalsIgnoreCase(trimmed)) {
                return attr;
            }
        }
        throw new IllegalArgumentException("Unknown card attribute: " + value);
    }
}

File: mobile/lib/domain/enums/card_attribute.dart
import 'package:freezed_annotation/freezed_annotation.dart';

/// Card elemental attributes.
@JsonEnum()
enum CardAttribute {
  @JsonValue('LIGHT')
  light('LIGHT'),
  @JsonValue('DARK')
  dark('DARK'),
  @JsonValue('WATER')
  water('WATER'),
  @JsonValue('FIRE')
  fire('FIRE'),
  @JsonValue('EARTH')
  earth('EARTH'),
  @JsonValue('WIND')
  wind('WIND'),
  @JsonValue('DIVINE')
  divine('DIVINE'),
  @JsonValue('UNKNOWN')
  unknown('UNKNOWN');

  final String value;
  const CardAttribute(this.value);

  static CardAttribute fromString(String? val) {
    if (val == null) return CardAttribute.unknown;
    return CardAttribute.values.firstWhere(
      (e) => e.value.toUpperCase() == val.toUpperCase(),
      orElse: () => CardAttribute.unknown,
    );
  }
}
